vmware-event-router/internal/provider/vcsim/vcsim.go

```diff
@@ -0,0 +1,187 @@
+
+	"github.com/pkg/errors"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/event"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
+
+	"github.com/vmware-samples/vcenter-event-broker-appliance/vmware-event-router/internal/color"
+	config "github.com/vmware-samples/vcenter-event-broker-appliance/vmware-event-router/internal/config/v1alpha1"
+	"github.com/vmware-samples/vcenter-event-broker-appliance/vmware-event-router/internal/events"
+	"github.com/vmware-samples/vcenter-event-broker-appliance/vmware-event-router/internal/metrics"
+	"github.com/vmware-samples/vcenter-event-broker-appliance/vmware-event-router/internal/processor"
+	"github.com/vmware-samples/vcenter-event-broker-appliance/vmware-event-router/internal/provider"
+)
+
+// EventStream handles the connection to the vCenter events API
+type EventStream struct {
+	client govmomi.Client
+	*log.Logger
+	verbose bool
+
+	sync.Mutex
+	stats metrics.EventStats
+}
+
+// assert we implement the provider interface
+var _ provider.Provider = (*EventStream)(nil)
+
+// NewEventStream returns a vCenter simulator event stream manager for a given
+// configuration and metrics server
+func NewEventStream(ctx context.Context, cfg *config.ProviderConfigVCSIM, ms metrics.Receiver, opts ...Option) (*EventStream, error) {
+	if cfg == nil {
+		return nil, errors.New("vCenter simulator configuration must be provided")
+	}
+
+	var vcsim EventStream
+
+	parsedURL, err := soap.ParseURL(cfg.Address)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing vCenter simulator URL")
+	}
+
+	// TODO: only supporting basic auth against vCenter simulator for now
+	if cfg.Auth == nil || cfg.Auth.BasicAuth == nil {
+		return nil, fmt.Errorf("invalid %s credentials: username and password must be set", config.BasicAuth)
+	}
+
+	username := cfg.Auth.BasicAuth.Username
+	password := cfg.Auth.BasicAuth.Password
+	parsedURL.User = url.UserPassword(username, password)
+
+	client, err := govmomi.NewClient(ctx, parsedURL, cfg.InsecureSSL)
+	if err != nil {
+		return nil, errors.Wrap(err, "create vCenter simulator client")
+	}
+
+	l := log.New(os.Stdout, color.Fata("[VCSIM] "), log.LstdFlags)
+	vcsim.Logger = l
+	vcsim.client = *client
+
+	// apply options (overwrite any defaults)
+	for _, opt := range opts {
+		opt(&vcsim)
+	}
+
+	// seed the metrics stats
+	vcsim.stats = metrics.EventStats{
+		Provider:    string(config.ProviderVCSIM),
+		Type:        config.EventProvider,
+		Address:     cfg.Address,
+		Started:     time.Now().UTC(),
+		EventsTotal: new(int),
+		EventsErr:   new(int),
+		EventsSec:   new(float64),
+	}
+
+	go vcsim.PushMetrics(ctx, ms)
+	return &vcsim, nil
+}
+
+// Stream implements the event provider interface and starts the event stream
+func (vcsim *EventStream) Stream(ctx context.Context, processor processor.Processor) error {
+	mgr := event.NewManager(vcsim.client.Client)
+	defer func() {
+		// ignore error against vcsim
+		_, _ = mgr.Destroy(ctx)
```

Comment:
I suppose because vcsim is a testing event provider, it's ok to ignore errors from `mgr.Destroy()`.

Reply:
I still like `defer mgr.Destroy(ctx)` over appeasing the linter when ignoring errors. But this is super minor, and it's all right to keep.

Reply:
Yeah, linter complains otherwise and pipeline would fail

Reply:
Ah, pipeline will fail. I see. Yeah, let's keep it, then.